File: pathplannerlib/src/main/java/com/pathplanner/lib/commands/PPMecanumControllerCommand.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package com.pathplanner.lib.commands;

import com.pathplanner.lib.PathPlannerTrajectory;
import com.pathplanner.lib.PathPlannerTrajectory.PathPlannerState;
import com.pathplanner.lib.controllers.PPHolonomicDriveController;
import com.pathplanner.lib.server.PathPlannerServer;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.MecanumDriveKinematics;
import edu.wpi.first.math.kinematics.MecanumDriveWheelSpeeds;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.*;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Supplier;

/** Custom PathPlanner version of MecanumControllerCommand */
public class PPMecanumControllerCommand extends CommandBase {
  private final Timer timer = new Timer();
  private final PathPlannerTrajectory trajectory;
  private final Supplier<Pose2d> poseSupplier;
  private final MecanumDriveKinematics kinematics;
  private final PPHolonomicDriveController controller;
  private final double maxWheelVelocityMetersPerSecond;
  private final Consumer<MecanumDriveWheelSpeeds> outputWheelSpeeds;
  private final Consumer<ChassisSpeeds> outputChassisSpeeds;
  private final boolean useKinematics;
  private final boolean useAllianceColor;

  private PathPlannerTrajectory transformedTrajectory;

  private static Consumer<PathPlannerTrajectory> logActiveTrajectory = null;
  private static Consumer<Pose2d> logTargetPose = null;
  private static Consumer<ChassisSpeeds> logSetpoint = null;
  private static BiConsumer<Translation2d, Rotation2d> logError =
      PPMecanumControllerCommand::defaultLogError;

  /**
   * Constructs a new PPMecanumControllerCommand that when executed will follow the provided
   * trajectory. The user should implement a velocity PID on the desired output wheel velocities.
   *
   * <p>Note: The controllers will *not* set the output to zero upon completion of the path - this
   * is left to the user, since it is not appropriate for paths with non-stationary end-states.
   *
   * @param trajectory The Pathplanner trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param xController The Trajectory Tracker PID controller for the robot's x position.
   * @param yController The Trajectory Tracker PID controller for the robot's y position.
   * @param rotationController The Trajectory Tracker PID controller for angle for the robot.
   * @param outputChassisSpeeds A consumer for a ChassisSpeeds object containing the output speeds.
   * @param useAllianceColor Should the path states be automatically transformed based on alliance
   *     color? In order for this to work properly, you MUST create your path on the blue side of
   *     the field.
   * @param requirements The subsystems to require.
   */
  public PPMecanumControllerCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      PIDController xController,
      PIDController yController,
      PIDController rotationController,
      Consumer<ChassisSpeeds> outputChassisSpeeds,
      boolean useAllianceColor,
      Subsystem... requirements) {
    this.trajectory = trajectory;
    this.poseSupplier = poseSupplier;
    this.kinematics = null;
    this.controller = new PPHolonomicDriveController(xController, yController, rotationController);
    this.maxWheelVelocityMetersPerSecond = 0;
    this.outputWheelSpeeds = null;
    this.outputChassisSpeeds = outputChassisSpeeds;
    this.useKinematics = false;
    this.useAllianceColor = useAllianceColor;

    addRequirements(requirements);

    if (useAllianceColor && trajectory.fromGUI && trajectory.getInitialPose().getX() > 8.27) {
      DriverStation.reportWarning(
          "You have constructed a path following command that will automatically transform path states depending"
              + " on the alliance color, however, it appears this path was created on the red side of the field"
              + " instead of the blue side. This is likely an error.",
          false);
    }
  }

  /**
   * Constructs a new PPMecanumControllerCommand that when executed will follow the provided
   * trajectory. The user should implement a velocity PID on the desired output wheel velocities.
   *
   * <p>Note: The controllers will *not* set the output to zero upon completion of the path - this
   * is left to the user, since it is not appropriate for paths with non-stationary end-states.
   *
   * @param trajectory The Pathplanner trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param xController The Trajectory Tracker PID controller for the robot's x position.
   * @param yController The Trajectory Tracker PID controller for the robot's y position.
   * @param rotationController The Trajectory Tracker PID controller for angle for the robot.
   * @param outputChassisSpeeds A consumer for a ChassisSpeeds object containing the output speeds.
   * @param requirements The subsystems to require.
   */
  public PPMecanumControllerCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      PIDController xController,
      PIDController yController,
      PIDController rotationController,
      Consumer<ChassisSpeeds> outputChassisSpeeds,
      Subsystem... requirements) {
    this(
        trajectory,
        poseSupplier,
        xController,
        yController,
        rotationController,
        outputChassisSpeeds,
        false,
        requirements);
  }

  /**
   * Constructs a new PPMecanumControllerCommand that when executed will follow the provided
   * trajectory. The user should implement a velocity PID on the desired output wheel velocities.
   *
   * <p>Note: The controllers will *not* set the output to zero upon completion of the path - this
   * is left to the user, since it is not appropriate for paths with non-stationary end-states.
   *
   * @param trajectory The Pathplanner trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param xController The Trajectory Tracker PID controller for the robot's x position.
   * @param yController The Trajectory Tracker PID controller for the robot's y position.
   * @param rotationController The Trajectory Tracker PID controller for angle for the robot.
   * @param maxWheelVelocityMetersPerSecond The maximum velocity of a drivetrain wheel.
   * @param outputWheelSpeeds A MecanumDriveWheelSpeeds object containing the output wheel speeds.
   * @param useAllianceColor Should the path states be automatically transformed based on alliance
   *     color? In order for this to work properly, you MUST create your path on the blue side of
   *     the field.
   * @param requirements The subsystems to require.
   */
  public PPMecanumControllerCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      MecanumDriveKinematics kinematics,
      PIDController xController,
      PIDController yController,
      PIDController rotationController,
      double maxWheelVelocityMetersPerSecond,
      Consumer<MecanumDriveWheelSpeeds> outputWheelSpeeds,
      boolean useAllianceColor,
      Subsystem... requirements) {
    this.trajectory = trajectory;
    this.poseSupplier = poseSupplier;
    this.kinematics = kinematics;
    this.controller = new PPHolonomicDriveController(xController, yController, rotationController);
    this.maxWheelVelocityMetersPerSecond = maxWheelVelocityMetersPerSecond;
    this.outputWheelSpeeds = outputWheelSpeeds;
    this.outputChassisSpeeds = null;
    this.useKinematics = true;
    this.useAllianceColor = useAllianceColor;

    addRequirements(requirements);

    if (useAllianceColor && trajectory.fromGUI && trajectory.getInitialPose().getX() > 8.27) {
      DriverStation.reportWarning(
          "You have constructed a path following command that will automatically transform path states depending"
              + " on the alliance color, however, it appears this path was created on the red side of the field"
              + " instead of the blue side. This is likely an error.",
          false);
    }
  }

  /**
   * Constructs a new PPMecanumControllerCommand that when executed will follow the provided
   * trajectory. The user should implement a velocity PID on the desired output wheel velocities.
   *
   * <p>Note: The controllers will *not* set the output to zero upon completion of the path - this
   * is left to the user, since it is not appropriate for paths with non-stationary end-states.
   *
   * @param trajectory The Pathplanner trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param xController The Trajectory Tracker PID controller for the robot's x position.
   * @param yController The Trajectory Tracker PID controller for the robot's y position.
   * @param rotationController The Trajectory Tracker PID controller for angle for the robot.
   * @param maxWheelVelocityMetersPerSecond The maximum velocity of a drivetrain wheel.
   * @param outputWheelSpeeds A MecanumDriveWheelSpeeds object containing the output wheel speeds.
   * @param requirements The subsystems to require.
   */
  public PPMecanumControllerCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      MecanumDriveKinematics kinematics,
      PIDController xController,
      PIDController yController,
      PIDController rotationController,
      double maxWheelVelocityMetersPerSecond,
      Consumer<MecanumDriveWheelSpeeds> outputWheelSpeeds,
      Subsystem... requirements) {
    this(
        trajectory,
        poseSupplier,
        kinematics,
        xController,
        yController,
        rotationController,
        maxWheelVelocityMetersPerSecond,
        outputWheelSpeeds,
        false,
        requirements);
  }

  @Override
  public void initialize() {
    if (useAllianceColor && trajectory.fromGUI) {
      transformedTrajectory =
          PathPlannerTrajectory.transformTrajectoryForAlliance(
              trajectory, DriverStation.getAlliance());
    } else {
      transformedTrajectory = trajectory;
    }

    if (logActiveTrajectory != null) {
      logActiveTrajectory.accept(transformedTrajectory);
    }

    this.timer.reset();
    this.timer.start();

    PathPlannerServer.sendActivePath(transformedTrajectory.getStates());
  }

  @Override
  public void execute() {
    double currentTime = this.timer.get();
    PathPlannerState desiredState = (PathPlannerState) transformedTrajectory.sample(currentTime);

    Pose2d currentPose = this.poseSupplier.get();

    PathPlannerServer.sendPathFollowingData(
        new Pose2d(desiredState.poseMeters.getTranslation(), desiredState.holonomicRotation),
        currentPose);

    ChassisSpeeds targetChassisSpeeds = this.controller.calculate(currentPose, desiredState);

    if (this.useKinematics) {
      MecanumDriveWheelSpeeds targetWheelSpeeds =
          this.kinematics.toWheelSpeeds(targetChassisSpeeds);

      targetWheelSpeeds.desaturate(this.maxWheelVelocityMetersPerSecond);

      this.outputWheelSpeeds.accept(targetWheelSpeeds);
    } else {
      this.outputChassisSpeeds.accept(targetChassisSpeeds);
    }

    if (logTargetPose != null) {
      logTargetPose.accept(
          new Pose2d(desiredState.poseMeters.getTranslation(), desiredState.holonomicRotation));
    }

    if (logError != null) {
      logError.accept(
          currentPose.getTranslation().minus(desiredState.poseMeters.getTranslation()),
          currentPose.getRotation().minus(desiredState.holonomicRotation));
    }

    if (logSetpoint != null) {
      logSetpoint.accept(targetChassisSpeeds);
    }
  }

  @Override
  public void end(boolean interrupted) {
    this.timer.stop();

    if (interrupted
        || Math.abs(transformedTrajectory.getEndState().velocityMetersPerSecond) < 0.1) {
      if (this.useKinematics) {
        this.outputWheelSpeeds.accept(new MecanumDriveWheelSpeeds(0, 0, 0, 0));
      } else {
        this.outputChassisSpeeds.accept(new ChassisSpeeds());
      }
    }
  }

  @Override
  public boolean isFinished() {
    return this.timer.hasElapsed(transformedTrajectory.getTotalTimeSeconds());
  }

  private static void defaultLogError(Translation2d translationError, Rotation2d rotationError) {
    SmartDashboard.putNumber("PPMecanumControllerCommand/xErrorMeters", translationError.getX());
    SmartDashboard.putNumber("PPMecanumControllerCommand/yErrorMeters", translationError.getY());
    SmartDashboard.putNumber(
        "PPMecanumControllerCommand/rotationErrorDegrees", rotationError.getDegrees());
  }

  /**
   * Set custom logging callbacks for this command to use instead of the default configuration of
   * pushing values to SmartDashboard
   *
   * @param logActiveTrajectory Consumer that accepts a PathPlannerTrajectory representing the
   *     active path. This will be called whenever a PPMecanumControllerCommand starts
   * @param logTargetPose Consumer that accepts a Pose2d representing the target pose while path
   *     following
   * @param logSetpoint Consumer that accepts a ChassisSpeeds object representing the setpoint
   *     speeds
   * @param logError BiConsumer that accepts a Translation2d and Rotation2d representing the error
   *     while path following
   */
  public static void setLoggingCallbacks(
      Consumer<PathPlannerTrajectory> logActiveTrajectory,
      Consumer<Pose2d> logTargetPose,
      Consumer<ChassisSpeeds> logSetpoint,
      BiConsumer<Translation2d, Rotation2d> logError) {
    PPMecanumControllerCommand.logActiveTrajectory = logActiveTrajectory;
    PPMecanumControllerCommand.logTargetPose = logTargetPose;
    PPMecanumControllerCommand.logSetpoint = logSetpoint;
    PPMecanumControllerCommand.logError = logError;
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/commands/PPMecanumControllerCommand.cpp
#include "pathplanner/lib/commands/PPMecanumControllerCommand.h"

#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/DriverStation.h>

using namespace pathplanner;

std::function<void(PathPlannerTrajectory)> PPMecanumControllerCommand::logActiveTrajectory =
		[](auto traj) {
		};
std::function<void(frc::Pose2d)> PPMecanumControllerCommand::logTargetPose = [](
		auto pose) {
};
std::function<void(frc::ChassisSpeeds)> PPMecanumControllerCommand::logSetpoint =
		[](auto speeds) {
		};
std::function<void(frc::Translation2d, frc::Rotation2d)> PPMecanumControllerCommand::logError =
		[](auto transError, auto rotError) {
			frc::SmartDashboard::PutNumber(
					"PPMecanumControllerCommand/xErrorMeters",
					transError.X()());
			frc::SmartDashboard::PutNumber(
					"PPMecanumControllerCommand/yErrorMeters",
					transError.Y()());
			frc::SmartDashboard::PutNumber(
					"PPMecanumControllerCommand/rotationErrorDegrees",
					rotError.Degrees()());
		};

PPMecanumControllerCommand::PPMecanumControllerCommand(
		PathPlannerTrajectory trajectory, std::function<frc::Pose2d()> pose,
		frc2::PIDController xController, frc2::PIDController yController,
		frc::PIDController thetaController,
		std::function<void(frc::ChassisSpeeds)> output,
		std::initializer_list<frc2::Subsystem*> requirements,
		bool useAllianceColor) : m_trajectory(std::move(trajectory)), m_pose(
		std::move(pose)), m_kinematics(frc::Translation2d(),
		frc::Translation2d(), frc::Translation2d(), frc::Translation2d()), m_controller(
		xController, yController, thetaController), m_maxWheelVelocity(0_mps), m_outputChassisSpeeds(
		output), m_useKinematics(false), m_useAllianceColor(useAllianceColor) {
	this->AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPMecanumControllerCommand::PPMecanumControllerCommand(
		PathPlannerTrajectory trajectory, std::function<frc::Pose2d()> pose,
		frc2::PIDController xController, frc2::PIDController yController,
		frc::PIDController thetaController,
		std::function<void(frc::ChassisSpeeds)> output,
		std::span<frc2::Subsystem* const > requirements, bool useAllianceColor) : m_trajectory(
		std::move(trajectory)), m_pose(std::move(pose)), m_kinematics(
		frc::Translation2d(), frc::Translation2d(), frc::Translation2d(),
		frc::Translation2d()), m_controller(xController, yController,
		thetaController), m_maxWheelVelocity(0_mps), m_outputChassisSpeeds(
		output), m_useKinematics(false), m_useAllianceColor(useAllianceColor) {
	this->AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPMecanumControllerCommand::PPMecanumControllerCommand(
		PathPlannerTrajectory trajectory, std::function<frc::Pose2d()> pose,
		frc::MecanumDriveKinematics kinematics, frc2::PIDController xController,
		frc2::PIDController yController, frc::PIDController thetaController,
		units::meters_per_second_t maxWheelVelocity,
		std::function<void(frc::MecanumDriveWheelSpeeds)> output,
		std::initializer_list<frc2::Subsystem*> requirements,
		bool useAllianceColor) : m_trajectory(std::move(trajectory)), m_pose(
		std::move(pose)), m_kinematics(kinematics), m_controller(xController,
		yController, thetaController), m_maxWheelVelocity(maxWheelVelocity), m_outputVel(
		output), m_useKinematics(true), m_useAllianceColor(useAllianceColor) {
	this->AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPMecanumControllerCommand::PPMecanumControllerCommand(
		PathPlannerTrajectory trajectory, std::function<frc::Pose2d()> pose,
		frc::MecanumDriveKinematics kinematics, frc2::PIDController xController,
		frc2::PIDController yController, frc::PIDController thetaController,
		units::meters_per_second_t maxWheelVelocity,
		std::function<void(frc::MecanumDriveWheelSpeeds)> output,
		std::span<frc2::Subsystem* const > requirements, bool useAllianceColor) : m_trajectory(
		std::move(trajectory)), m_pose(std::move(pose)), m_kinematics(
		kinematics), m_controller(xController, yController, thetaController), m_maxWheelVelocity(
		maxWheelVelocity), m_outputVel(output), m_useKinematics(true), m_useAllianceColor(
		useAllianceColor) {
	this->AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

void PPMecanumControllerCommand::Initialize() {
	if (m_useAllianceColor && m_trajectory.fromGUI) {
		m_transformedTrajectory =
				PathPlannerTrajectory::transformTrajectoryForAlliance(
						m_trajectory, frc::DriverStation::GetAlliance());
	} else {
		m_transformedTrajectory = m_trajectory;
	}

	m_timer.Reset();
	m_timer.Start();

	if (PPMecanumControllerCommand::logActiveTrajectory) {
		PPMecanumControllerCommand::logActiveTrajectory (m_transformedTrajectory);
	}
}

void PPMecanumControllerCommand::Execute() {
	auto currentTime = m_timer.Get();
	auto desiredState = m_transformedTrajectory.sample(currentTime);

	frc::Pose2d currentPose = m_pose();

	auto targetChassisSpeeds = m_controller.calculate(currentPose,
			desiredState);

	if (m_useKinematics) {
		auto targetWheelSpeeds = m_kinematics.ToWheelSpeeds(
				targetChassisSpeeds);

		targetWheelSpeeds.Desaturate(m_maxWheelVelocity);

		m_outputVel(targetWheelSpeeds);
	} else {
		m_outputChassisSpeeds(targetChassisSpeeds);
	}

	if (PPMecanumControllerCommand::logTargetPose) {
		PPMecanumControllerCommand::logTargetPose(
				frc::Pose2d(desiredState.pose.Translation(),
						desiredState.holonomicRotation));
	}

	if (PPMecanumControllerCommand::logError) {
		PPMecanumControllerCommand::logError(
				currentPose.Translation() - desiredState.pose.Translation(),
				currentPose.Rotation() - desiredState.holonomicRotation);
	}

	if (PPMecanumControllerCommand::logSetpoint) {
		PPMecanumControllerCommand::logSetpoint(targetChassisSpeeds);
	}
}

void PPMecanumControllerCommand::End(bool interrupted) {
	m_timer.Stop();

	if (interrupted
			|| std::abs(m_transformedTrajectory.getEndState().velocity())
					< 0.1) {
		if (m_useKinematics) {
			m_outputVel(frc::MecanumDriveWheelSpeeds());
		} else {
			m_outputChassisSpeeds(frc::ChassisSpeeds());
		}
	}
}

bool PPMecanumControllerCommand::IsFinished() {
	return m_timer.HasElapsed(m_transformedTrajectory.getTotalTime());
}
